runtime/compiler/optimizer/HandleRecompilationOps.cpp

```diff
@@ -0,0 +1,258 @@
+#include "il/TreeTop.hpp"
+#include "il/TreeTop_inlines.hpp"
+#include "infra/Cfg.hpp"
+#include "infra/CfgEdge.hpp"
+#include "optimizer/HandleRecompilationOps.hpp"
+#include "optimizer/Optimization.hpp"
+#include "optimizer/Optimization_inlines.hpp"
+
+int32_t
+TR_HandleRecompilationOps::perform()
+   {
+   _enableTransform = true;
+
+   if (trace())
+      {
+      traceMsg(comp(), "Entering HandleRecompilationOps\n");
+      }
+
+   // The following series of tests check whether voluntary OSR is allowed and that recompilation is allowed.
+   // If not, _enableTransform is set to false, but the optimizer still checks for situations in which it
+   // would like to induce OSR.  If any of those situations are encountered with _enableTransform set to
+   // false, the compilation will abort, as correct code generation for those situations requires OSR to be
+   // induced.
+   //
+   if (comp()->getOption(TR_DisableOSR))
+      {
+      if (trace())
+         {
+         traceMsg(comp(), "Disabling Handle Recompilation Operations as OSR is disabled\n");
+         }
+
+      _enableTransform = false;
+      }
+
+   if (comp()->getHCRMode() != TR::osr)
+      {
+      if (trace())
+         {
+         traceMsg(comp(), "Disabling Handle Recompilation Operations as HCR mode is not OSR\n");
+         }
+
+      _enableTransform = false;
+      }
+
+   if (comp()->getOSRMode() == TR::involuntaryOSR)
+      {
+      if (trace())
+         {
+         traceMsg(comp(), "Disabling Handle Recompilation Operations as OSR mode is involuntary\n");
+         }
+
+      _enableTransform = false;
+      }
+
+   if (!comp()->supportsInduceOSR())
+      {
+      if (trace())
+         {
+         traceMsg(comp(), "Disabling Handle Recompilation Operations as induceOSR is not supported\n");
+         }
+
+      _enableTransform = false;
+      }
+
+   if (!comp()->allowRecompilation())
+      {
+      if (trace())
+         {
+         traceMsg(comp(), "Disabling Handle Recompilation Operations as recompilation is not permitted\n");
+         }
+
+      _enableTransform = false;
+      }
+
+   for (TR::TreeTop *tt = comp()->getStartTree(); tt != NULL; tt = tt->getNextTreeTop())
+      {
+      TR::Node *node = tt->getNode();
+      visitNode(tt, node);
+      }
+
+   if (_enableTransform && trace())
+     {
+     traceMsg(comp(), "Exiting HandleRecompilationOps\n");
+     }
+
+   return 0;
+   }
+
+bool
+TR_HandleRecompilationOps::resolveCHKGuardsValueTypeOperation(TR::TreeTop *currTree, TR::Node *node)
+   {
+   TR::Node *resolveChild = node->getFirstChild();
+
+   if (resolveChild->getOpCodeValue() == TR::loadaddr)
+      {
+      // Check whether the ResolveCHK guards a loadaddr that is used in newvalue
+      // operation.
+      //
+      TR::SymbolReference *loadAddrSymRef = resolveChild->getSymbolReference();
+
+      if (loadAddrSymRef->isUnresolved())
+         {
+         for (TR::TreeTop *nextTree = currTree->getNextTreeTop();
```

Comment:
You might want to look at the TreeTop iterator in the walkers to save some boilerplate

Reply:
I neglected to address this comment.  I'll take care of this next.

Reply:
Addressed in commit 305c99e431ebe5ac8995080f366479e4df9e43b6